services/MerchStoreService.ts

```diff
@@ -410,40 +414,176 @@ export default class MerchStoreService {
+   * the user who placed the order can only cancel the order, while the admin can mark
+   * it as missed.
+   *
+   * If the pickup event of the order is changed, then it must be changed to a pickup event that's
+   * at least 2 days in the future.
+   *
+   * @param uuid order uuid
+   * @param orderEdit order edits
+   * @param user user requesting the edit
+   * @returns the edited order
+   */
+  public async editOrder(uuid: Uuid, orderEdit: MerchOrderEdit, user: UserModel): Promise<OrderModel> {
+    return this.transactions.readWrite(async (txn) => {
+      const orderRespository = Repositories.merchOrder(txn);
+      const order = await orderRespository.findByUuid(uuid);
+      if (!order) throw new NotFoundError('Order not found');
+      if (!user.isAdmin() && order.user !== user) throw new ForbiddenError('Member cannot cancel other members orders');
+
+      const { status, pickupEvent: pickupEventUuid } = orderEdit;
+      if (status) {
+        const upsertedOrder = await orderRespository.upsertMerchOrder(order, { status });
+        const orderUpdateInfo = await MerchStoreService.buildOrderUpdateInfo(order, txn);
+        switch (status) {
+          case OrderStatus.CANCELLED:
+            if (new Date() > moment(order.pickupEvent.start).subtract(2, 'days').toDate()) {
+              throw new NotFoundError('Cannot cancel an order with a pickup date less than 2 days away');
+            }
+            await MerchStoreService.refundUser(user, order.totalCost, txn);
+            await this.emailService.sendOrderCancellation(user.email, user.firstName, orderUpdateInfo);
+            break;
+          case OrderStatus.PICKUP_MISSED:
+            await this.emailService.sendOrderPickupMissed(user.email, user.firstName, orderUpdateInfo);
+            break;
+          default:
+            // status can only be CANCELLED or PICKUP_MISSED due to controller validation, so default case can be empty
+            break;
+        }
+        return upsertedOrder;
+      }
+      // verify the requested pickup event exists,
+      // and that the order is placed at least 2 days before the pickup event starts
+      const pickupEvent = await Repositories.merchOrderPickupEvent(txn).findByUuid(pickupEventUuid);
+      if (!pickupEvent) {
+        throw new NotFoundError('Pickup event requested is not found');
+      }
+      if (new Date() > moment(pickupEvent.start).subtract(2, 'days').toDate()) {
+        throw new NotFoundError('Cannot pickup order at an event that starts in less than 2 days');
+      }
+      return orderRespository.upsertMerchOrder(order, { pickupEvent });
+    });
+  }
+
+  /**
+   * Maps an item's option to its price at purchase and quantity ordered by the user
+   * @param order order
+   * @returns map of item option to its price at purchase and quantity ordered by the user
+   */
+  private static getOptionPriceAndQuantitiesFromOrder(order: OrderModel): Map<string, OrderItemPriceAndQuantity> {
```

Comment:
Nit: if we're working with maps we should include the key in the name.
```suggestion
  private static getPriceAndQuantityByOption(order: OrderModel): Map<Uuid, OrderItemPriceAndQuantity> {
```